nodejs/week3/src/backend/api/meals.js

```diff
@@ -0,0 +1,148 @@
+const express = require("express");
+const app = express();
+const router = express.Router();
+const knex = require("../database");
+
+//Get all meals
+router.get("/", async (req, res) => {
+  try {
+    const mealslist = await knex("meals").select("*");
+    res.send(mealslist);
+  } catch (error) {
+    throw error;
+  }
+});
+
+router.get("/titles", async (request, response) => {
+  try {
+    // knex syntax for selecting table records.
+    const titles = await knex("meals").select("title");
+    response.json(titles);
+  } catch (error) {
+    throw error;
+  }
+});
+
+//GET api/meals/ query parameters
+
+router.get("/", async function (req, res) {
+  const {
+    maxPrice,
+    title,
+    availableReservations,
+    createdAfter,
+    limit,
+  } = req.query;
+  const mealTitles = await knex.select("*").table("meals");
+
+  if (Object.keys(req.query).length === 0) {
+    res.send(mealTitles);
+  }
+
+  if (maxPrice) {
+    try {
+      const maxPriceMeals = await knex(meals)
+        .select("*")
+        .where("price", "<", maxPrice);
+      console.log(maxPriceMeals);
+      res.send(maxPriceMeals);
+    } catch (error) {
+      error;
+    }
+  } else if (title) {
+    try {
+      const titles = await knex(meals)
+        .select("*")
+        .where("title", "like", `%${title}%`);
+      console.log(titles);
+      res.send(titles);
+    } catch (error) {
+      error;
+    }
+  } else if (availableReservations === "true") {
+    try {
+      const availableReservationsMeal = await knex
+        .select("*")
+        .sum({
+          sumTotal: "reservations.number_of_guests",
+        })
+        .table("meals")
+        .join("reservations", {
+          "meals.id": "reservations.meal_id",
+        })
+        .groupBy("meals.title")
+        .having("sumTotal", ">", "Max_reservations");
+      console.log(availableReservationsMeal);
+      res.send(availableReservationsMeal);
+    } catch (error) {
+      error;
+    }
+  } else if (createdAfter) {
+    try {
+      const datecreated = await knex
+        .select("*")
+        .table("meals")
+        .where("created_date", "<", createdAfter);
+      console.log(datecreated);
+      res.send(datecreated);
+    } catch (error) {
+      error;
+    }
+  } else if (limit) {
+    try {
+      const limitMeals = await knex.select("*").table("meals").limit(limit);
+      console.log(limitMeals);
+      res.send(limitMeals);
+    } catch (error) {
+      error;
+    }
+  }
+});
+
+//Adds a new meal
+router.post("/", async (req, res) => {
+  const meal = {
+    id: req.body.id,
```

Comment:
what happens if the database already has a meal with this id?
what happens if the request provides wrong data? e.g. the price is not a number?

Reply:
A good advice is to never trust data provided by a user (or a client/consumer) of an API or application. You need to validate any type of data that you receive from the outside world so that it doesn't break your application in unexpected ways. If an error occurs you want to communicate to the client of your application what the error is and how they can recover from that error. That helps you troubleshoot issues and guides your development.

(e.g. if I get a 404 not found I know that the id that I'm sending is probably wrong, if I get a 400 bad request I knoow that there's some parameter in my request that is wrong, etc)